include/libp2p/security/noise/crypto/state.hpp

```diff
@@ -0,0 +1,223 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef LIBP2P_INCLUDE_LIBP2P_SECURITY_NOISE_STATE_HPP
+#define LIBP2P_INCLUDE_LIBP2P_SECURITY_NOISE_STATE_HPP
+
+#include <boost/optional.hpp>
+
+#include <libp2p/security/noise/crypto/interfaces.hpp>
+#include <libp2p/security/noise/crypto/message_patterns.hpp>
+
+namespace libp2p::security::noise {
+
+  enum class Error {
+    INTERNAL_ERROR = 1,
+    WRONG_KEY32_SIZE,
+    EMPTY_HANDSHAKE_NAME,
+    WRONG_PRESHARED_KEY_SIZE,
+    NOT_INITIALIZED,
+    UNEXPECTED_WRITE_CALL,
+    UNEXPECTED_READ_CALL,
+    NO_HANDSHAKE_MESSAGE,
+    MESSAGE_TOO_LONG,
+    MESSAGE_TOO_SHORT,
+    NO_PUBLIC_KEY,
+    REMOTE_KEY_ALREADY_SET,
+  };
+
+  outcome::result<Key32> bytesToKey32(gsl::span<const uint8_t> key);
+
+  /// Provides symmetric encryption and decryption after a successful handshake
+  class CipherState {
```

Comment:
I'm not sure if I completely understand why does a Cipher_State_ encrypts and decrypts something and I believe that it would give me hard time understanding what it is without the walkthrough through the PR.

Reply:
Agree. The names was preserved from go implementation. Moving encrypt/decrypt methods out of "states" will require significant redesign. Left it as is for now.